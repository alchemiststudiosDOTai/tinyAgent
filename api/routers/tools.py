from fastapi import APIRouter, HTTPException
from core.agent import Agent
from ..schemas.models import ToolResult
from core.tools import (
    brave_web_search_tool,
    duckduckgo_web_search,
    anon_coder_tool,
    llm_serializer_tool,
    ripgrep_tool,
    aider_tool,
    file_manipulator_tool,
    custom_text_browser_tool,
    final_answer_extractor
)
from core.mcp import ensure_mcp_server
import logging
import os
from core import tool

@tool
def clean_response_tool(messy_response: str) -> str:
    """Clean and format a messy response using LLM to make it human-readable.
    
    Args:
        messy_response: The raw, messy response that needs cleaning
        
    Returns:
        A clean, formatted response suitable for human reading
    """
    # Create an agent for cleaning
    agent = Agent()
    
    # Use the agent's run method with a specific prompt
    prompt = f"""Please clean up and format this response to make it human-readable. 
    Remove any technical metadata, debug information, or formatting characters.
    Keep only the meaningful content and present it in a clear, organized way.
    
    Messy response:
    {messy_response}
    
    Clean response:"""
    
    # Use the agent's run method instead of accessing llm directly
    return agent.run(query=prompt, template_path="workflows/triage.md")

router = APIRouter()
logger = logging.getLogger(__name__)

@router.post("/chat", response_model=ToolResult)
async def chat_endpoint(message: str, session_id: str = None):
    try:
        # Create triage agent
        agent = Agent()
        
        # Ensure MCP server is running
        mcp_available = ensure_mcp_server()
        if not mcp_available:
            logger.warning("MCP server is not available, some tools may not work")
        
        # Register all available tools that have the correct interface
        tools_to_register = [
            anon_coder_tool,
            llm_serializer_tool,
            ripgrep_tool,
            aider_tool,
            file_manipulator_tool,
            custom_text_browser_tool,
            final_answer_extractor,
            clean_response_tool  # Add the new response cleaner tool
        ]
        
        # Only add MCP-dependent tools if MCP server is available
        if mcp_available:
            tools_to_register.extend([brave_web_search_tool, duckduckgo_web_search])
            logger.info("MCP server is available, MCP-dependent tools loaded")
        
        for tool in tools_to_register:
            if hasattr(tool, 'name') and hasattr(tool, 'description') and hasattr(tool, 'func') and hasattr(tool, 'parameters'):
                agent.create_tool(tool.name, tool.description, tool.func, tool.parameters)
            else:
                logger.warning(f"Skipping tool registration for {tool.__class__.__name__} - missing required attributes")
        
        # Send message to triage agent using the triage template
        raw_response = agent.run(
            query=message,
            template_path="workflows/triage.md"
        )
        
        # Clean the response using the clean_response_tool
        cleaned_response = clean_response_tool(raw_response)
        
        # Return cleaned response
        return ToolResult(
            tool_name="chat",
            result=cleaned_response,
            success=True
        )
    except Exception as e:
        logger.error(f"Chat error: {str(e)}")
        raise HTTPException(500, "Internal server error")
