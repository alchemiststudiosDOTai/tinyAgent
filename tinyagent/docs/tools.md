# tinyAgent Tools Framework

This document provides a comprehensive guide to the tools framework in tinyAgent, including how to create, use, and extend tools.

## What Are Tools?

In tinyAgent, tools are the functional units that the Agent can execute to accomplish tasks. Each tool:

1. Has a unique name
2. Has a descriptive explanation of what it does
3. Defines parameters it accepts and their types
4. Implements a specific function
5. Returns a result that can be used by the Agent

Tools are the primary way to extend tinyAgent's capabilities, allowing it to perform a wide range of tasks from simple calculations to complex API interactions.

## Tool Types

tinyAgent supports several types of tools:

1. **Python Tools** - Written directly in Python using the `Tool` class or `@tool` decorator
2. **External Tools** - Written in other languages (Go, Rust, C/C++, etc.) with JSON communication
3. **MCP Tools** - Provided by MCP servers, extending capabilities dynamically

## Creating Python Tools

### Using the Tool Class

The most direct way to create a tool is by instantiating the `Tool` class:

```python
from core import Tool, ParamType

def calculate_sum(a: int, b: int) -> int:
    """Calculate the sum of two integers."""
    return a + b

sum_tool = Tool(
    name="calculate_sum",
    description="Calculate the sum of two integers",
    parameters={
        "a": ParamType.INTEGER,
        "b": ParamType.INTEGER
    },
    func=calculate_sum
)
```

### Using the @tool Decorator

For a more concise approach, you can use the `@tool` decorator:

```python
from core import tool

@tool
def calculate_sum(a: int, b: int) -> int:
    """Calculate the sum of two integers."""
    return a + b

# The decorator automatically creates a Tool instance
# based on the function's signature and docstring
```

The decorator can also take arguments to customize the tool:

```python
@tool(name="add_numbers", rate_limit=5)
def calculate_sum(a: int, b: int) -> int:
    """Calculate the sum of two integers."""
    return a + b
```

## Parameter Types

tinyAgent supports the following parameter types via the `ParamType` enum:

- `ParamType.STRING` - For text values
- `ParamType.INTEGER` - For whole numbers
- `ParamType.FLOAT` - For decimal numbers
- `ParamType.ANY` - For any type of value

When using the `@tool` decorator, parameter types are inferred from Python type hints:

```python
@tool
def format_name(first: str, last: str, age: int) -> str:
    """Format a person's name and age."""
    return f"{first} {last} is {age} years old"
# Equivalent to:
# parameters={"first": ParamType.STRING, "last": ParamType.STRING, "age": ParamType.INTEGER}
```

## Tool Factory and Registration

Tools are now managed through the `AgentFactory`, which provides centralized tool registration, execution, and tracking:

```python
from core.factory.agent_factory import AgentFactory
from core.agent import Agent

# Get the singleton factory instance
factory = AgentFactory.get_instance()

# Create and register a tool
factory.create_tool(
    name="calculator",
    description="Perform basic arithmetic operations",
    func=calculate
)

# Create an agent that will use the factory
agent = Agent(factory=factory)

# Run the agent with a query
result = agent.run("Calculate 5 + 3")
print(result)  # Output: 8
```

This approach ensures proper tool tracking and rate limiting. The factory automatically:

1. Registers tools with unique names
2. Tracks usage statistics for each tool
3. Enforces rate limits
4. Provides a centralized execution path

### Direct Tool Registration

You can also register existing Tool instances with the factory:

```python
# Create a tool
my_tool = Tool(
    name="my_tool",
    description="My custom tool",
    parameters={"param": ParamType.STRING},
    func=my_function
)

# Register with the factory
factory.register_tool(my_tool)
```

## Tool Execution Flow

All tool executions now go through the factory to ensure consistent tracking and rate limiting:

```python
# Bad: Direct tool invocation (bypasses tracking & rate limiting)
# result = my_tool(param="value")  # Don't do this!

# Good: Factory-mediated execution
result = factory.execute_tool("my_tool", param="value")
```

When using an Agent, tool execution is automatically handled through the factory:

```python
# Agent.run will use factory.execute_tool internally
result = agent.run("Use my_tool with value")
```

## Creating Cross-Language Tools

One of tinyAgent's most powerful features is the ability to incorporate tools written in other programming languages through a standardized JSON interface.

### JSON Communication Protocol

External tools communicate with tinyAgent through a simple JSON protocol:

1. **Input**: The tool receives a JSON object through stdin containing the tool parameters
2. **Processing**: The tool processes the input and generates a result
3. **Output**: The tool returns a JSON object through stdout containing the result

### Manifest File

External tools require a `manifest.json` file that defines the tool's metadata:

```json
{
  "name": "text_rank",
  "description": "Extract key sentences from text using TextRank algorithm",
  "parameters": {
    "text": {
      "type": "string",
      "description": "The text to analyze"
    },
    "num_sentences": {
      "type": "integer",
      "description": "Number of key sentences to extract",
      "default": 3
    }
  },
  "command": "./text_rank",
  "working_directory": "./external_tools/text_rank/"
}
```

### Example: C Implementation

Here's an example of a tool implemented in C:

```c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <json-c/json.h>

int main() {
    // Read JSON from stdin
    char buffer[4096];
    fgets(buffer, sizeof(buffer), stdin);
    
    // Parse input JSON
    struct json_object *input_json = json_tokener_parse(buffer);
    
    // Extract parameters
    struct json_object *text_obj, *num_sentences_obj;
    json_object_object_get_ex(input_json, "text", &text_obj);
    json_object_object_get_ex(input_json, "num_sentences", &num_sentences_obj);
    
    const char *text = json_object_get_string(text_obj);
    int num_sentences = json_object_get_int(num_sentences_obj);
    
    // Process text using your algorithm
    // ... (implementation of text processing)
    
    // Create result JSON
    struct json_object *result_json = json_object_new_object();
    struct json_object *sentences_array = json_object_new_array();
    
    // Add extracted sentences to array
    for (int i = 0; i < num_sentences; i++) {
        // Example - in real code, you'd use your algorithm's results
        char sentence[100];
        sprintf(sentence, "This is extracted sentence %d", i+1);
        json_object_array_add(sentences_array, json_object_new_string(sentence));
    }
    
    json_object_object_add(result_json, "sentences", sentences_array);
    
    // Output result JSON
    printf("%s\n", json_object_to_json_string(result_json));
    
    // Cleanup
    json_object_put(input_json);
    json_object_put(result_json);
    
    return 0;
}
```

### Example: Rust Implementation

For performance-critical tools, languages like Rust offer excellent speed and safety:

```rust
use serde::{Deserialize, Serialize};
use std::io::{self, Read};

#[derive(Deserialize)]
struct Input {
    text: String,
    num_sentences: Option<usize>,
}

#[derive(Serialize)]
struct Output {
    sentences: Vec<String>,
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Read JSON from stdin
    let mut buffer = String::new();
    io::stdin().read_to_string(&mut buffer)?;
    
    // Parse input
    let input: Input = serde_json::from_str(&buffer)?;
    let num_sentences = input.num_sentences.unwrap_or(3);
    
    // Process text - this would use your actual algorithm
    let sentences = extract_key_sentences(&input.text, num_sentences);
    
    // Create and output result
    let output = Output { sentences };
    println!("{}", serde_json::to_string(&output)?);
    
    Ok(())
}

fn extract_key_sentences(text: &str, count: usize) -> Vec<String> {
    // This is where your TextRank or other algorithm would be implemented
    // Just a placeholder implementation for the example
    text.split('.')
        .take(count)
        .map(|s| s.trim().to_string())
        .filter(|s| !s.is_empty())
        .collect()
}
```

### Loading External Tools

External tools are loaded using the `load_external_tools` function:

```python
from core.tools import load_external_tools

# Load external tools from a directory
external_tools = load_external_tools("/path/to/tools")

# Register with the factory
factory = AgentFactory.get_instance()
for tool in external_tools:
    factory.register_tool(tool)
```

## Built-in Tools

tinyAgent includes several built-in tools that provide common functionality:

### anon_coder

Executes Python code safely with configurable security checks:

```python
from core.tools import anon_coder_tool

# Direct usage through factory
result = factory.execute_tool(
    "anon_coder",
    code="print([x**2 for x in range(10)])",
    timeout=5
)
```

Security settings for code execution can be configured in `config.yml`:

```yaml
# Code execution security configuration
code_execution:
  # Allow potentially dangerous operations (file operations, etc)
  allow_dangerous_operations: false  # Set to true to disable security restrictions
  
  # Optional: more granular control
  allowed_operations:
    file_operations: false
    os_operations: false
    imports: ["os", "sys"]  # Additional allowed imports beyond the defaults
```

This allows for balancing security with functionality based on your specific needs and trust level.

### llm_serializer

Uses LLMs to convert complex objects to JSON-compatible formats:

```python
from core.tools import llm_serializer_tool

# Convert a complex object to JSON via factory
result = factory.execute_tool(
    "llm_serializer",
    obj_type="CustomClass",
    obj_repr=str(my_complex_object)
)
```

### ripgrep

Searches files using the ripgrep tool:

```python
# Search for a pattern in files via factory
result = factory.execute_tool(
    "ripgrep",
    pattern="TODO",
    path="src/",
    flags="-i"
)
```

## Tool Error Handling

Tools should handle errors gracefully and return informative error messages:

```python
@tool
def divide(a: float, b: float) -> float:
    """Divide two numbers."""
    try:
        if b == 0:
            raise ValueError("Cannot divide by zero")
        return a / b
    except Exception as e:
        raise ToolError(f"Division failed: {str(e)}")
```

The Agent class will catch and handle exceptions raised by tools, including:

- `ToolError` - General tool execution errors
- `ToolNotFoundError` - When a tool is not found
- `ToolExecutionError` - When a tool execution fails
- `RateLimitExceeded` - When a tool's rate limit is exceeded

## Rate Limiting

Tools can have rate limits to prevent excessive usage:

```python
# Tool class with rate limit
api_tool = Tool(
    name="api_call",
    description="Call an external API",
    parameters={"query": ParamType.STRING},
    func=make_api_call,
    rate_limit=10  # Max 10 calls per session
)

# Decorator with rate limit
@tool(rate_limit=5)
def limited_function(param: str) -> str:
    """A rate-limited function."""
    return f"Processed: {param}"

# Factory with rate limit
factory.create_tool(
    name="expensive_api",
    description="Call an expensive API",
    func=call_expensive_api,
    rate_limit=5
)
```

Rate limits can also be configured globally in `config.yml`:

```yaml
# Rate limiting configuration
rate_limits:
  global_limit: 30  # Default limit for all tools
  tool_limits:      # Tool-specific limits
    web_search: 10
    code_execution: 5
    api_call: 20
```

## Monitoring Tool Usage

The factory provides methods to monitor tool usage:

```python
# Get current status
status = factory.get_status()

print("\nTool usage statistics:")
for tool_name, stats in status["tools"].items():
    print(f"  {tool_name}: {stats['calls']}/{stats['limit']} calls")
```

## Best Practices

### Naming

Tool names should be:
- Lowercase
- No spaces (use underscores)
- Descriptive of what the tool does
- Concise

### Descriptions

Tool descriptions should:
- Clearly explain what the tool does
- Include examples of usage
- Specify any requirements or limitations
- Be detailed enough for the LLM to understand when to use it

### Parameters

Parameter design should:
- Use clear, descriptive names
- Use the appropriate type for each parameter
- Include only necessary parameters
- Use default values where appropriate

### Implementation

Tool implementation should:
- Handle errors gracefully with informative messages
- Validate inputs before processing
- Be efficient and avoid unnecessary computation
- Clean up any resources (files, connections) it uses

### Cross-Language Tools

When creating tools in other languages:
- Use standard JSON libraries for your language
- Implement proper error handling and report errors in the JSON response
- Keep the tool focused on a single capability
- Consider performance characteristics (startup time, memory usage)
- Document any dependencies or installation requirements

## Conclusion

Tools are the core extension mechanism for tinyAgent. By creating and using tools, you can extend tinyAgent's capabilities to handle a wide range of tasks. 

The flexibility of the tools framework allows for tools to be written in Python or other languages, making it easy to leverage:

1. High-performance libraries in languages like Rust or C++
2. Existing code bases in any language
3. Language-specific capabilities (e.g., GPU programming, system-level operations)
4. Specialized algorithms with better implementations in other languages

The standardized JSON interface creates a universal language for tool communication, enabling truly polyglot agent systems while maintaining a simple, consistent approach to tool usage.
