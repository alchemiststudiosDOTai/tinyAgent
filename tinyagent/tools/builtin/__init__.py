"""tinyagent.tools.builtin package exports."""

from .planning import create_plan, get_plan, update_plan
from .web_search import web_search

__all__ = [
    "create_plan",
    "get_plan",
    "update_plan",
    "web_search",
]
