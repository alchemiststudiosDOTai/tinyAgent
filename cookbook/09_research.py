#!/usr/bin/env python3
"""
Enhanced Business Research Tool with Advanced Proxy Support

Features:
1. Multiple proxy support (Tor, rotating proxies, direct)
2. Advanced error handling and retries
3. Configurable search parameters
4. Enhanced data storage
5. Rate limiting protection
"""

import os
import sys
import json
import time
import random
import urllib.request
from typing import Optional, Dict, List, Generator
from datetime import datetime
from pathlib import Path
from dataclasses import dataclass
from dotenv import load_dotenv

from duckduckgo_search import DDGS
from duckduckgo_search.exceptions import (
    DuckDuckGoSearchException,
    RatelimitException,
    TimeoutException,
)

# Add parent directory to the path so we can import the core package
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from core.factory.agent_factory import AgentFactory
from core.tools.duckduckgo_search import duckduckgo_search_tool, perform_duckduckgo_search

@dataclass
class ProxyConfig:
    """Proxy configuration settings."""
    enabled: bool
    url_template: str
    username: Optional[str] = None
    password: Optional[str] = None
    country: str = "US"
    
    @classmethod
    def from_config(cls) -> 'ProxyConfig':
        """Create proxy config from config.yml."""
        load_dotenv()
        
        try:
            import yaml
            with open('config.yml', 'r') as f:
                config = yaml.safe_load(f)
            
            proxy_config = config.get('proxy', {})
            
            if not proxy_config.get('enabled', False):
                return cls(enabled=False, url_template="")
            
            return cls(
                enabled=True,
                url_template=proxy_config.get('url', ''),
                username=os.getenv('TINYAGENT_PROXY_USERNAME'),
                password=os.getenv('TINYAGENT_PROXY_PASSWORD'),
                country=os.getenv('TINYAGENT_PROXY_COUNTRY', 'US')
            )
        except Exception as e:
            print(f"Warning: Failed to load proxy config: {str(e)}")
            return cls(enabled=False, url_template="")
    
    def get_proxy_string(self) -> Optional[str]:
        """Get formatted proxy string using the URL template."""
        if not self.enabled or not self.url_template:
            return None
            
        if not all([self.username, self.password]):
            print("Warning: Proxy credentials not found in environment variables")
            return None
            
        try:
            # Format the URL template with credentials
            proxy_url = self.url_template % (
                self.username,
                self.country,
                self.password
            )
            return proxy_url
        except Exception as e:
            print(f"Warning: Failed to format proxy URL: {str(e)}")
            return None

    def get_proxy_handler(self) -> Optional[urllib.request.ProxyHandler]:
        """Get configured proxy handler for urllib."""
        proxy_url = self.get_proxy_string()
        if not proxy_url:
            return None
        
        return urllib.request.ProxyHandler({
            'http': proxy_url,
            'https': proxy_url
        })

class ResearchManager:
    """Manages research operations with proxy support and error handling."""
    
    def __init__(self, proxy_config: ProxyConfig):
        self.proxy_config = proxy_config
        self.ddgs = None
        self.last_rotation = time.time()
        self.initialize_ddgs()
    
    def initialize_ddgs(self):
        """Initialize or reinitialize the DDGS client."""
        proxy = self.proxy_config.get_proxy_string()
        self.ddgs = DDGS(
            proxy=proxy,
            timeout=20
        )
        self.last_rotation = time.time()
    
    def should_rotate_proxy(self) -> bool:
        """Check if it's time to rotate the proxy."""
        if not self.proxy_config.enabled:
            return False
        # Rotate every 60 seconds if proxy is enabled
        return time.time() - self.last_rotation >= 60

    def search_with_retries(
        self,
        keywords: str,
        region: str = "wt-wt",
        safesearch: str = "moderate",
        timelimit: Optional[str] = None,
        backend: str = "auto",
        max_results: Optional[int] = None,
        max_retries: int = 3,
        retry_delay: int = 5
    ) -> List[Dict[str, str]]:
        """
        Perform search with automatic retries and proxy rotation.
        
        Args:
            keywords: Search query
            region: Search region (e.g., 'wt-wt', 'us-en')
            safesearch: SafeSearch setting ('on', 'moderate', 'off')
            timelimit: Time limit for results ('d', 'w', 'm', 'y')
            backend: Search backend ('auto', 'html', 'lite')
            max_results: Maximum number of results to return
            max_retries: Maximum number of retry attempts
            retry_delay: Delay between retries in seconds
            
        Returns:
            List of search results
        """
        attempts = 0
        while attempts < max_retries:
            try:
                if self.should_rotate_proxy():
                    self.initialize_ddgs()
                
                results = list(self.ddgs.text(
                    keywords=keywords,
                    region=region,
                    safesearch=safesearch,
                    timelimit=timelimit,
                    backend=backend,
                    max_results=max_results
                ))
                
                return results
                
            except RatelimitException:
                print(f"Rate limit hit on attempt {attempts + 1}/{max_retries}")
                if self.proxy_config.enabled:
                    self.initialize_ddgs()
                time.sleep(retry_delay * (attempts + 1))
                
            except TimeoutException:
                print(f"Timeout on attempt {attempts + 1}/{max_retries}")
                time.sleep(retry_delay)
                
            except DuckDuckGoSearchException as e:
                print(f"Search error on attempt {attempts + 1}/{max_retries}: {str(e)}")
                if attempts == max_retries - 1:
                    raise
                time.sleep(retry_delay)
                
            attempts += 1
        
        raise DuckDuckGoSearchException(f"Failed after {max_retries} attempts")

def setup_output_directory() -> Path:
    """Create and return the output directory path."""
    output_dir = Path("output/research_results")
    output_dir.mkdir(parents=True, exist_ok=True)
    return output_dir

def save_results(results: List[Dict[str, str]], query: str, aspect: str, output_dir: Path) -> str:
    """
    Save search results with enhanced metadata.
    
    Args:
        results: List of search results
        query: Search query used
        aspect: Research aspect
        output_dir: Output directory path
        
    Returns:
        Path to saved file
    """
    timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
    filename = f"research_{aspect}_{timestamp}.json"
    filepath = output_dir / filename
    
    data = {
        "metadata": {
            "timestamp": datetime.now().isoformat(),
            "query": query,
            "aspect": aspect,
            "result_count": len(results)
        },
        "results": results
    }
    
    with open(filepath, 'w', encoding='utf-8') as f:
        json.dump(data, f, indent=2, ensure_ascii=False)
    
    return str(filepath)

def enhance_research_query(base_query: str, aspect: str = "") -> str:
    """
    Enhance search query with aspect-specific terms.
    
    Args:
        base_query: Base search query
        aspect: Research aspect to focus on
        
    Returns:
        Enhanced query string
    """
    research_aspects = {
        "substance": "chemical composition effects uses research studies",
        "market": "market size vendors suppliers distribution statistics analysis",
        "legal": "legal status regulation compliance requirements FDA policy",
        "safety": "safety studies side effects research clinical trials risks",
        "production": "manufacturing process quality control standards certification"
    }
    
    if aspect and aspect in research_aspects:
        return f"{base_query} {research_aspects[aspect]}"
    return base_query

def format_search_results(results: list) -> None:
    """Format and print search results in a readable way."""
    if not results:
        print("No results found")
        return

    print("\n=== Search Results ===")
    for i, result in enumerate(results, 1):
        print(f"\nResult {i}:")
        print("-" * 50)
        print(f"Title: {result.get('title', 'N/A')}")
        print(f"URL: {result.get('url', 'N/A')}")
        print("\nSnippet:")
        print(result.get('snippet', 'No snippet available'))
        print("-" * 50)

def main():
    """Run the enhanced research tool."""
    # Setup
    output_dir = setup_output_directory()
    proxy_config = ProxyConfig.from_config()
    research_mgr = ResearchManager(proxy_config)
    
    # Research configuration
    research_config = {
        "substance": {
            "query": "starts ups in USA",
            "max_results": 10
        },
     
    }
    
    print("Enhanced Research Tool")
    print("====================")
    print(f"Proxy enabled: {proxy_config.enabled}")
    if proxy_config.enabled:
        print(f"Using proxy with country: {proxy_config.country}")
    print(f"Output directory: {output_dir}")
    print()
    
    all_results = {}
    
    # Run research for each aspect
    for aspect, config in research_config.items():
        print(f"\nResearching {aspect.upper()}")
        print("-" * 50)
        
        try:
            # Enhance the query with aspect-specific terms
            enhanced_query = enhance_research_query(config["query"], aspect)
            print(f"Enhanced query: {enhanced_query}")
            
            # Perform the search with retries and proxy rotation
            results = research_mgr.search_with_retries(
                keywords=enhanced_query,
                max_results=config["max_results"],
                region="us-en",
                safesearch="moderate",
                timelimit="y",
                backend="auto"
            )
            
            # Save results
            output_file = save_results(
                results=results,
                query=enhanced_query,
                aspect=aspect,
                output_dir=output_dir
            )
            
            print(f"Found {len(results)} results")
            print(f"Saved to: {output_file}")
            
            all_results[aspect] = {
                "query": enhanced_query,
                "results": results,
                "output_file": output_file
            }
            
        except DuckDuckGoSearchException as e:
            print(f"Error researching {aspect}: {str(e)}")
            continue
        
        except Exception as e:
            print(f"Unexpected error researching {aspect}: {str(e)}")
            continue
    
    # Print summary
    print("\nResearch Summary")
    print("===============")
    for aspect, data in all_results.items():
        print(f"\n{aspect.upper()}:")
        print(f"- Results: {len(data['results'])}")
        print(f"- Output: {data['output_file']}")

if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        print("\nResearch interrupted by user")
    except Exception as e:
        print(f"Fatal error: {str(e)}")
        sys.exit(1)
