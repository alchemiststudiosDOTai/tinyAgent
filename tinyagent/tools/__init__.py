"""tinyagent.tools package exports."""

from .builtin import web_search
from .validation import ToolValidationError, validate_tool_class

__all__ = [
    "web_search",
    "validate_tool_class",
    "ToolValidationError",
]
