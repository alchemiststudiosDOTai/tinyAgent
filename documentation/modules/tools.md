# Tools Documentation

The `tinyagent` toolkit is built around a lightweight registry that turns plain Python callables into ReAct-compatible tools. This page explains how the registry works, how to register new tools, and what built-ins ship with the library.

## Overview

- `tinyagent.tool` is a decorator that registers a callable with the global tool registry.
- Registered functions are wrapped in the `Tool` dataclass, capturing the original function, signature, and docstring.
- `get_registry()` exposes a read-only snapshot of the active tools.
- `freeze_registry()` locks the registry to prevent accidental mutation at runtime.

## Quick Start

```python
from tinyagent import ReactAgent, tool

@tool
def calculator(expression: str) -> float:
    """Evaluate a mathematical expression."""
    return eval(expression, {"__builtins__": {}}, {})

agent = ReactAgent(tools=[calculator])
result = agent.run("What is 12 * 7?")
```

The decorator records the function under its name (`"calculator"` in this example). Docstrings become the tool description that the agent includes in the system prompt, so keep them concise and actionable.

## What the Decorator Does

`tool` is an alias for `ToolRegistry.register`. When you decorate a function:

1. A `Tool` wrapper is created with:
   - `fn`: original callable
   - `name`: the function name
   - `doc`: stripped docstring
   - `signature`: `inspect.signature(fn)` for argument binding
2. The wrapper is stored in the global registry under `fn.__name__`.
3. The original function is returned untouched, so you can still call it directly or pass it to an agent.

The `Tool` wrapper exposes two execution paths:

- `tool_fn(*args, **kwargs)` → calls the original function.
- `tool_fn.run(payload)` → binds a JSON payload (`{"arg": "value"}`) to the signature and returns the stringified result. This is how `ReactAgent` executes tools after parsing LLM output.

## Managing the Registry

```python
from tinyagent import freeze_registry, get_registry

registry_view = get_registry()  # Mapping[str, Tool]
for name, tool_obj in registry_view.items():
    print(name, tool_obj.doc)

freeze_registry()  # further registrations raise RuntimeError
```

- Use `get_registry()` when you want to introspect available tools (e.g., for debugging or documentation). The returned mapping is read-only.
- Call `freeze_registry()` after wiring your tools to guarantee no new registrations happen later in the run. Attempts to add or mutate tools after freezing raise `RuntimeError`.

## Built-in Tools

`tinyagent.tools.web_search` is the only built-in tool today. It integrates with the Brave Search API and requires a `BRAVE_SEARCH_API_KEY` environment variable.

```python
import os
from tinyagent.tools import web_search

os.environ["BRAVE_SEARCH_API_KEY"] = "sk-your-token"
print(web_search("latest reinforcement learning papers"))
```

Errors are returned as strings (e.g., missing API key, network failures), so agents can surface them as observations.

## Tool Class Validation

Use `tinyagent.validate_tool_class` (aliased at `tinyagent.tools.validate_tool_class`) to lint class-based tools before exposing them to the agent:

```python
from tinyagent import ToolValidationError, validate_tool_class

class WeatherTool:
    """Fetch the forecast for a location."""

    api = "https://weather.example/api"

    def __init__(self, units: str = "metric") -> None:
        self.units = units

    def run(self, city: str) -> str:
        ...

validate_tool_class(WeatherTool)  # raises ToolValidationError on violations
```

The validator performs static checks using the Python AST:

- Class attributes must be literals (str, int, dict, etc.).
- `__init__` parameters must all provide literal default values; no `*args`/`**kwargs`.
- Methods cannot reference undefined names or rely on implicit globals.

Violations are reported in a single `ToolValidationError`, making it easy to enforce safe, serializable tooling.

## Best Practices

- **Type hints:** Always annotate tool parameters for clarity and better OpenAPI-style prompts.
- **Docstrings:** Write purposeful docstrings—these become tool descriptions in the agent prompt.
- **Argument binding:** Ensure parameter names are simple (`snake_case`) because the agent must match them exactly in its JSON calls.
- **Freezing:** Freeze the registry in long-lived processes to avoid accidental re-registration.
- **Testing:** Exercise tools directly in unit tests; treat `tool_fn.run(payload)` as the interface the agent will use.
