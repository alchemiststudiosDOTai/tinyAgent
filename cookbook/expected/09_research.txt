Using model: deepseek/deepseek-r1
Max steps per research phase: 2
Debug level: 0
Configuration source: defaults

Research phases:
  - gathering: 2 steps
  - analysis: 2 steps
  - synthesis: 2 steps
Initializing custom text browser...
=== CONFIGURING PROXY ===
Proxy disabled in configuration
Proxy disabled for browser
Initializing file manager...
Research output directory: tinyAgent_output/research
Initializing tools registry...
Registered 8 gathering tools, 5 analysis tools, and 3 synthesis tools

Initialization complete

Using model: deepseek/deepseek-r1
Max steps per research phase: 2
Debug level: 0
Rate limit delay: 2.0s
Max retries: 3
Retry delay: 1.0s
Proxy configured: False

=== PERFORM_DUCKDUCKGO_SEARCH DEBUG ===
Search parameters: keywords=research about the latest trends general_analysis latest developments research, max_results=5, region=wt-wt, safesearch=moderate, timelimit=None, backend=auto
Using proxy: http://customer-could_sxa0C-cc-US:********@pr.oxylabs.io:7777
Found 5 results
=== END PERFORM_DUCKDUCKGO_SEARCH DEBUG ===


=== PERFORM_DUCKDUCKGO_SEARCH DEBUG ===
Search parameters: keywords=research about the latest trends technical latest developments research, max_results=5, region=wt-wt, safesearch=moderate, timelimit=None, backend=auto
Using proxy: http://customer-could_sxa0C-cc-US:********@pr.oxylabs.io:7777
Found 5 results
=== END PERFORM_DUCKDUCKGO_SEARCH DEBUG ===


=== PERFORM_DUCKDUCKGO_SEARCH DEBUG ===
Search parameters: keywords=research about the latest trends implementation latest developments research, max_results=5, region=wt-wt, safesearch=moderate, timelimit=None, backend=auto
Using proxy: http://customer-could_sxa0C-cc-US:********@pr.oxylabs.io:7777
Found 5 results
=== END PERFORM_DUCKDUCKGO_SEARCH DEBUG ===

